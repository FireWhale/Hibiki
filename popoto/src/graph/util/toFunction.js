/**
 * Convert the parameter to a function returning the parameter if it is not already a function.
 *
 * @param param
 * @return {*}
 */

export default function toFunction(param) {
    if (typeof param === "function") {
        return param;
    } else {
        return function () {
            return param;
        };
    }
}

