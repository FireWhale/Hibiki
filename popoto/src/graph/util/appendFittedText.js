import * as d3 from "d3";
import toFunction from "./toFunction";

var CONTEXT_2D = document.createElement("canvas").getContext("2d");
var DEFAULT_CANVAS_LINE_HEIGHT = 12;

export function measureTextWidth(text) {
    return CONTEXT_2D.measureText(text).width;
}

/**
 * Compute the radius of the circle wrapping all the lines.
 *
 * @param lines array of text lines
 * @return {number}
 */
function computeTextRadius(lines) {
    var textRadius = 0;

    for (var i = 0, n = lines.length; i < n; ++i) {
        var dx = lines[i].width / 2;
        var dy = (Math.abs(i - n / 2 + 0.5) + 0.5) * DEFAULT_CANVAS_LINE_HEIGHT;
        textRadius = Math.max(textRadius, Math.sqrt(dx * dx + dy * dy));
    }

    return textRadius;
}

function computeLines(words, targetWidth) {
    var line;
    var lineWidth0 = Infinity;
    var lines = [];

    for (var i = 0, n = words.length; i < n; ++i) {
        var lineText1 = (line ? line.text + " " : "") + words[i];
        var lineWidth1 = measureTextWidth(lineText1);
        if ((lineWidth0 + lineWidth1) / 2 < targetWidth) {
            line.width = lineWidth0 = lineWidth1;
            line.text = lineText1;
        } else {
            lineWidth0 = measureTextWidth(words[i]);
            line = {width: lineWidth0, text: words[i]};
            lines.push(line);
        }
    }

    return lines;
}

function computeTargetWidth(text) {
    return Math.sqrt(measureTextWidth(text.trim()) * DEFAULT_CANVAS_LINE_HEIGHT);
}

/**
 * Split text into words.
 *
 * @param text
 * @return {*|string[]}
 */
function computeWords(text) {
    var words = text.split(/\s+/g); // To hyphenate: /\s+|(?<=-)/
    if (!words[words.length - 1]) words.pop();
    if (!words[0]) words.shift();
    return words;
}

/**
 * Inspired by https://beta.observablehq.com/@mbostock/fit-text-to-circle
 * Extract words from the text and group them by lines to fit a circle.
 *
 * @param text
 * @returns {*}
 */
export function extractLines(text) {
    if (text === undefined || text === null) {
        return [];
    }

    var textString = String(text);

    var words = computeWords(textString);
    var targetWidth = computeTargetWidth(textString);

    return computeLines(words, targetWidth);
}

function createTextElements(selection, getClass) {
    selection.each(function (d) {
        var text = d3.select(this)
            .selectAll(".fitted-text")
            .data([{}]);

        text.enter()
            .append("text")
            .merge(text)
            .attr("class", "fitted-text" + (getClass !== undefined ? " " + getClass(d) : ""))
            .attr("style", "text-anchor: middle; font: 10px sans-serif");
    });
}

function createSpanElements(text, getText) {
    text.each(function (fitData) {
        var lines = extractLines(getText(fitData));
        var span = d3.select(this).selectAll("tspan")
            .data(lines);

        span.exit().remove();

        span.enter()
            .append("tspan")
            .merge(span)
            .attr("x", 0)
            .attr("y", function (d, i) {
                var lineCount = lines.length;
                return (i - lineCount / 2 + 0.8) * DEFAULT_CANVAS_LINE_HEIGHT
            })
            .text(function (d) {
                return d.text
            });
    });
}

/**
 * Create the text representation of a node by slicing the text into lines to fit the node.
 *
 * @param selection
 * @param textParam
 * @param radiusParam
 * @param classParam
 */
export default function appendFittedText(selection, textParam, radiusParam, classParam) {
    var getRadius = toFunction(radiusParam);
    var getText = toFunction(textParam);
    var getClass = classParam ? toFunction(classParam) : classParam;

    createTextElements(selection, getClass);

    var text = selection.select(".fitted-text");

    createSpanElements(text, getText);

    text.attr("transform", function (d) {
        var lines = extractLines(getText(d));
        var textRadius = computeTextRadius(lines);

        var scale = 1;
        if (textRadius !== 0 && textRadius) {
            scale = getRadius(d) / textRadius;
        }
        return "translate(" + 0 + "," + 0 + ")" + " scale(" + scale + ")"
    });
}