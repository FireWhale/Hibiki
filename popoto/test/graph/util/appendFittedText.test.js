import * as d3 from "d3"
import {default as appendFittedText, measureTextWidth, extractLines} from '../../../src/graph/util/appendFittedText'

describe("measureTextWidth", function () {
    test("Should measure undefined text correctly", () => {
        expect(measureTextWidth()).toMatchSnapshot();
    });

    test("Should measure empty text correctly", () => {
        expect(measureTextWidth("")).toMatchSnapshot();
    });

    test("Should measure text correctly", () => {
        expect(measureTextWidth("mockedTextValue")).toMatchSnapshot();
    });
});

describe("extractLines", function () {

    test("Should works with undefined as text", () => {
        const lines = extractLines();
        expect(lines).toMatchSnapshot();
    });

    test("Should works with empty string as text", () => {
        const lines = extractLines("");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with simple word as text", () => {
        const lines = extractLines("mockedTextValue");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple words as text", () => {
        const lines = extractLines("mocked text value");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple words on same line", () => {
        const lines = extractLines("mocked text a value");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with multiple spaces and multiple words as text", () => {
        const lines = extractLines("        mocked   text       value           ");
        expect(lines).toMatchSnapshot();
    });

    test("Should works with integer as text", () => {
        const lines = extractLines(0);
        expect(lines).toMatchSnapshot();
    });

});

describe("appendFittedText", function () {

    describe('Should works with text, radius, class parameters as value without data in selection', () => {
        beforeEach(() => {
            const textMockValue = "mocked_text_value";
            const radiusMockValue = 50;
            const cssClassMockValue = "mocked_css_class_value";

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '</div>';
            appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should works without class parameters', () => {
        beforeEach(() => {
            const textMockValue = "mocked_text_value";
            const radiusMockValue = 50;

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '</div>';
            appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should works with text, radius, class parameters as function without data in selection', () => {
        beforeEach(() => {
            const textMockValue = function (d) {
                return "mocked_text_value"
            };
            const radiusMockValue = function (d) {
                return 50
            };
            const cssClassMockValue = function (d) {
                return "mocked_css_class_value"
            };

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '</div>';

            appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should works with text, radius, class parameters as function with data in selection', () => {
        beforeEach(() => {
            const textMockValue = function (d) {
                return d.textMockValue
            };
            const radiusMockValue = function (d) {
                return d.radiusMockValue
            };
            const cssClassMockValue = function (d) {
                return d.cssClassMockValue
            };

            const data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
            ];

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '<g>' +
                '</g>' +
                '</div>';

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue)
        });

        test('Should render correctly', () => {
            expect(document.body.innerHTML).toMatchSnapshot();
        });
    });

    describe('Should not mutate data', () => {
        const data = [
            {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
            {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
        ];

        beforeEach(() => {
            const textMockValue = function (d) {
                return d.textMockValue
            };
            const radiusMockValue = function (d) {
                return d.radiusMockValue
            };
            const cssClassMockValue = function (d) {
                return d.cssClassMockValue
            };

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '<g>' +
                '</g>' +
                '</div>';

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
        });

        describe('Should works with text, radius, class parameters as value without data in selection', () => {
            beforeEach(() => {
                const textMockValue = "mocked_text_value";
                const radiusMockValue = 50;
                const cssClassMockValue = "mocked_css_class_value";

                document.body.innerHTML =
                    '<div>' +
                    '<g>' +
                    '</g>' +
                    '</div>';
                appendFittedText(d3.selectAll('g').data([{}]), textMockValue, radiusMockValue, cssClassMockValue)
            });

            test('Should render correctly', () => {
                expect(document.body.innerHTML).toMatchSnapshot();
            });
        });

    });

    describe('Should update correctly', () => {
        let textMockValue;
        let radiusMockValue;
        let cssClassMockValue;
        beforeEach(() => {
            textMockValue = function (d) {
                return d.textMockValue
            };
            radiusMockValue = function (d) {
                return d.radiusMockValue
            };
            cssClassMockValue = function (d) {
                return d.cssClassMockValue
            };

            document.body.innerHTML =
                '<div>' +
                '<g>' +
                '</g>' +
                '<g>' +
                '</g>' +
                '</div>';

        });

        test('Should update correctly if data changed', () => {
            let data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

            data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value3", radiusMockValue: 53, cssClassMockValue: "mocked_css_class_value3"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

        });

        test('Should update correctly if data add text div', () => {
            let data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

            data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

        });

        test('Should update correctly if data add span', () => {
            let data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked_text_value2", radiusMockValue: 52, cssClassMockValue: "mocked_css_class_value2"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

            data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"},
                {textMockValue: "mocked text value 3", radiusMockValue: 53, cssClassMockValue: "mocked_css_class_value3"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

        });

        test('Should update correctly if data remove span', () => {

            let data = [
                {textMockValue: "mocked text value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

            data = [
                {textMockValue: "mocked_text_value", radiusMockValue: 50, cssClassMockValue: "mocked_css_class_value"}
            ];

            appendFittedText(d3.selectAll('g').data(data), textMockValue, radiusMockValue, cssClassMockValue);
            expect(document.body.innerHTML).toMatchSnapshot();

        });
    });
});