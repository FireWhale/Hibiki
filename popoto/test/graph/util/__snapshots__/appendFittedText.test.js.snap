// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appendFittedText Should not mutate data Should works with text, radius, class parameters as value without data in selection Should render correctly 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data add span 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value2\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48074017006186526)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value2</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data add span 2`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value3\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(1.2249629333307142)\\"><tspan x=\\"0\\" y=\\"-14.399999999999999\\">mocked</tspan><tspan x=\\"0\\" y=\\"-2.3999999999999995\\">text</tspan><tspan x=\\"0\\" y=\\"9.600000000000001\\">value</tspan><tspan x=\\"0\\" y=\\"21.6\\">3</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data add text div 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data add text div 2`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value2\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48074017006186526)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value2</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data changed 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value2\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48074017006186526)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value2</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data changed 2`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value3\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48998517333228575)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value3</tspan></text></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data remove span 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(1.2422599874998832)\\"><tspan x=\\"0\\" y=\\"-8.399999999999999\\">mocked</tspan><tspan x=\\"0\\" y=\\"3.6000000000000005\\">text</tspan><tspan x=\\"0\\" y=\\"15.600000000000001\\">value</tspan></text></g><g></g></div>"`;

exports[`appendFittedText Should update correctly Should update correctly if data remove span 2`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g></g></div>"`;

exports[`appendFittedText Should works with text, radius, class parameters as function with data in selection Should render correctly 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g><g><text class=\\"fitted-text mocked_css_class_value2\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.48074017006186526)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value2</tspan></text></g></div>"`;

exports[`appendFittedText Should works with text, radius, class parameters as function without data in selection Should render correctly 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`appendFittedText Should works with text, radius, class parameters as value without data in selection Should render correctly 1`] = `"<div><g><text class=\\"fitted-text mocked_css_class_value\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`appendFittedText Should works without class parameters Should render correctly 1`] = `"<div><g><text class=\\"fitted-text\\" style=\\"text-anchor: middle; font: 10px sans-serif\\" transform=\\"translate(0,0) scale(0.4893501829289196)\\"><tspan x=\\"0\\" y=\\"3.6000000000000005\\">mocked_text_value</tspan></text></g></div>"`;

exports[`extractLines Should works with empty string as text 1`] = `Array []`;

exports[`extractLines Should works with integer as text 1`] = `
Array [
  Object {
    "text": "0",
    "width": 12,
  },
]
`;

exports[`extractLines Should works with multiple spaces and multiple words as text 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "value",
    "width": 60,
  },
]
`;

exports[`extractLines Should works with multiple words as text 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "value",
    "width": 60,
  },
]
`;

exports[`extractLines Should works with multiple words on same line 1`] = `
Array [
  Object {
    "text": "mocked",
    "width": 72,
  },
  Object {
    "text": "text",
    "width": 48,
  },
  Object {
    "text": "a value",
    "width": 84,
  },
]
`;

exports[`extractLines Should works with simple word as text 1`] = `
Array [
  Object {
    "text": "mockedTextValue",
    "width": 180,
  },
]
`;

exports[`extractLines Should works with undefined as text 1`] = `Array []`;

exports[`measureTextWidth Should measure empty text correctly 1`] = `0`;

exports[`measureTextWidth Should measure text correctly 1`] = `180`;

exports[`measureTextWidth Should measure undefined text correctly 1`] = `0`;
